04_condition_variable/exercises/03a_ping_pong.cpp

```diff
@@ -1,26 +1,92 @@
+#include <chrono>
+#include <condition_variable>
+#include <cstdlib>
 #include <iostream>
+#include <mutex>
+#include <sstream>
 #include <thread>
-#include <chrono>
+
 using namespace std;
 
 class PingPong {
     int repetitions_;
+    int repetitionsCounter_;
+    mutex coutMutex_;
+    condition_variable pingOrPong_;
+    bool pingTurn_;
 
 public:
     PingPong(int repetitions)
-        : repetitions_(repetitions)
+        : repetitions_(repetitions), repetitionsCounter_(1), pingTurn_(false)
     {}
 
+    void safePongExit(stringstream& ss) {
+        lock_guard<mutex> lockCout(coutMutex_);
+        ss << "Repetitions end!\n";
+        cout << ss.rdbuf();
+        exit(0);
+    }
+    
+    void safeStopExit() {
+        lock_guard<mutex> lockCout(coutMutex_);
+        stringstream ss;
+        
+        if(pingTurn_ == false) {
+            ss << "pong" << repetitionsCounter_ - 1 << '\n';
+            cout << ss.rdbuf();
+        }
+
+        ss << "Timeout!\n";
+        cout << ss.rdbuf();
+        exit(0);
+    }
```

Comment:
ta funkcja niby spełnia założenia formalne, ale czasowe tak nie do końca. Jak minie timeout to od razu kończy i tylko dopisze brakującego ponga od razu, a więc nie będzie on po tym swoim własnym timeoutcie. 
`exit(0)` też trochę traktuję jak taki hack. A co jeśli taka funkcjonalność typu ping-pong była jedną z wielu, które robi program i nie chcielibyśmy go ubijać tylko np miałby po zatrzymaniu robić coś innego? 

Ja bym sugerował w poczekać dopóki pong sam się nie wypisze. Niektórzy robili tak, że pingTurn nie był boolem, tylko enumem i mógł przyjąć 3 wartości - ping, pong, timeout :)

Reply:
Najprościej jak się da spełniając wymagania ;) Ale rozumiem o co Ci chodzi. Pomysł z enumem dobry, wykorzystam go.